src/main/scala/org/ergoplatform/modifiers/mempool/ErgoTransaction.scala

```diff
@@ -170,71 +177,88 @@ case class ErgoTransaction(override val inputs: IndexedSeq[Input],
       .validateSeq(boxesToSpend.zipWithIndex) { case (validation, (box, idx)) =>
         val currentTxCost = validation.result.payload.get
 
-        val input = inputs(idx)
-        val proof = input.spendingProof
-        val transactionContext = TransactionContext(boxesToSpend, dataBoxes, this)
-        val inputContext = InputContext(idx.toShort, proof.extension)
-
-        val ctx = new ErgoContext(
-          stateContext, transactionContext, inputContext,
-          costLimit = maxCost - currentTxCost, // remaining cost so far
-          initCost = 0)
-
-        val costTry = verifier.verify(box.ergoTree, ctx, proof, messageToSign)
-        costTry.recover { case t =>
-          log.debug(s"Tx verification failed: ${t.getMessage}")
+        val (res, time) = measureTimeNano {
+          val input = inputs(idx)
+          val proof = input.spendingProof
+          val transactionContext = TransactionContext(boxesToSpend, dataBoxes, this)
+          val inputContext = InputContext(idx.toShort, proof.extension)
+
+          val ctx = new ErgoContext(
+            stateContext, transactionContext, inputContext,
+            costLimit = maxCost - currentTxCost, // remaining cost so far
+            initCost = 0)
+
+          val costTry = verifier.verify(box.ergoTree, ctx, proof, messageToSign)
+          costTry.recover { case t =>
+            log.debug(s"Tx verification failed: ${t.getMessage}")
+          }
+
+          val (isCostValid, scriptCost) = costTry.getOrElse((false, maxCost + 1))
+
+          val currCost = addExact(currentTxCost, scriptCost)
+
+          validation
+            // Just in case, should always be true if client implementation is correct.
+            .validateEquals(txBoxToSpend, box.id, input.boxId)
+            // Check whether input box script interpreter raised exception
+            .validate(txScriptValidation, costTry.isSuccess && isCostValid, s"$id: #$idx => $costTry")
+            // Check that cost of the transaction after checking the input becomes too big
+            .validate(bsBlockTransactionsCost, currCost <= maxCost, s"$id: cost exceeds limit after input #$idx")
+            .map(c => addExact(c, scriptCost))
         }
 
-        val (isCostValid, scriptCost) = costTry.getOrElse((false, maxCost + 1))
-
-        val currCost = addExact(currentTxCost, scriptCost)
-
-        validation
-          // Just in case, should always be true if client implementation is correct.
-          .validateEquals(txBoxToSpend, box.id, input.boxId)
-          // Check whether input box script interpreter raised exception
-          .validate(txScriptValidation, costTry.isSuccess && isCostValid, s"$id: #$idx => $costTry")
-          // Check that cost of the transaction after checking the input becomes too big
-          .validate(bsBlockTransactionsCost, currCost <= maxCost, s"$id: cost exceeds limit after input #$idx")
-          .map(c => addExact(c, scriptCost))
-    }
+        collectMetricsTo(
```

Comment:
Extract as method is lengthy and complex  even without the new lines (also, consensus-critical)